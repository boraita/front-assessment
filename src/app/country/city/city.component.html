<div class="container">
  <div class="capital-container">
    <div
      class="select-country"
      *ngIf="(getCountry() | async) == null; else capital"
    >
      Select a country
    </div>
    <ng-template #capital>
      <div class="capital-selected">
        <div class="country-name">
          <span> Country </span>
          <span>{{ (getCountry() | async)?.country }}</span>
        </div>
        <div class="city-name">
          <span> Capital </span>
          <span> {{ (getCountry() | async)?.city || "Unknown" }} </span>
        </div>
      </div>
    </ng-template>
  </div>
</div>
