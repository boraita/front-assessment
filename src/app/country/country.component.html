<div class="main">
    <app-country-list class="country-list"></app-country-list>
    <app-city></app-city>
</div>
  