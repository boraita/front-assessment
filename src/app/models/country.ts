export type Country = {
    country: string;
    city: string;
}