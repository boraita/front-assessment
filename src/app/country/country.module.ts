import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CountryListComponent } from './country-list/country-list.component';
import { ShowCityComponent } from './show-city/show-city.component';
import { HttpClientModule } from '@angular/common/http';
import { CountryService } from './services/country.service';
import { CountryComponent } from './country.component';
import { RouterModule } from '@angular/router';

@NgModule({
  declarations: [CountryListComponent, ShowCityComponent, CountryComponent],
  imports: [
    CommonModule,
    HttpClientModule,
    RouterModule.forChild([
      {
        path: '',
        component: CountryComponent,
      },
    ]),
  ],
  providers: [CountryService],
})
export class CountryModule {}
