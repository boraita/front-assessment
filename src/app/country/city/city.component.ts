import { Component, Input } from '@angular/core';
import { Country } from '../../models/country';
import { CountryService } from '../services/country.service';
import { Observable } from 'rxjs';

@Component({
  selector: 'app-city',
  templateUrl: './city.component.html',
  styleUrls: ['./city.component.css'],
})
export class ShowCityComponent {
  constructor(private readonly countryService: CountryService) {}

  getCountry(): Observable<Country | null> {
    return this.countryService.getSelectedCountry();
  }
}
