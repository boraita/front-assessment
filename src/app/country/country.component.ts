import { Component } from '@angular/core';
import { Country } from '../models/country';

@Component({
  selector: 'app-country',
  templateUrl: './country.component.html',
  styleUrls: ['./country.component.css'],
})
export class CountryComponent {
  country!: Country;
}
