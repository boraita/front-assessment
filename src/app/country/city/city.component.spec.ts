import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { of } from 'rxjs';
import { Country } from '../../models/country';
import { CountryService } from '../services/country.service';
import { ShowCityComponent } from './city.component';

jest.mock('./../services/country.service');

describe('ShowCityComponent', () => {
  let component: ShowCityComponent;
  let fixture: ComponentFixture<ShowCityComponent>;
  let countryService: CountryService;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ShowCityComponent],
      providers: [CountryService],
    }).compileComponents();

    fixture = TestBed.createComponent(ShowCityComponent);
    component = fixture.componentInstance;
    countryService = TestBed.inject(CountryService);

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('country should be undefined', () => {
    expect(component.getCountry()).toBeUndefined();
    const divSelectCity = fixture.debugElement.query(By.css('.select-country'));
    expect(divSelectCity.nativeElement.textContent.trim()).toBe(
      'Select a country'
    );
  });

  it('should be shown capital Unknown', () => {
    jest
      .spyOn(countryService, 'getSelectedCountry')
      .mockReturnValue(of({ country: 'Afghanistan', city: null }));
    fixture.detectChanges();
    const divSelectedCountry = fixture.debugElement.query(By.css('.city-name'));
    expect(divSelectedCountry.nativeElement.textContent.trim()).toBe(
      'Capital  Unknown'
    );
  });

  it('should be shown capital from country selected', () => {
    jest.spyOn(countryService, 'getSelectedCountry').mockReturnValue(
      of({
        country: 'Spain',
        city: 'Madrid',
      } as Country)
    );
    fixture.detectChanges();
    const divSelectedCountry = fixture.debugElement.query(By.css('.city-name'));
    expect(divSelectedCountry.nativeElement.textContent.trim()).toBe(
      'Capital  Madrid'
    );
  });

  it('should display Select a country (Snapshot)', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should display a capital from input country value (Snapshot)', () => {
    jest.spyOn(countryService, 'getSelectedCountry').mockReturnValue(
      of({
        country: 'Spain',
        city: 'Madrid',
      } as Country)
    );
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
