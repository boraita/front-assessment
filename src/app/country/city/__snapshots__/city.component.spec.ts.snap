// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowCityComponent should display Select a country (Snapshot) 1`] = `
<app-city
  __ngContext__={[Function LRootView]}
  countryService={[Function CountryService]}
>
  <div
    class="container"
  >
    <div
      class="capital-container"
    >
      <div
        class="select-country"
      >
         Select a country 
      </div>
    </div>
  </div>
</app-city>
`;

exports[`ShowCityComponent should display Select a country 1`] = `
<app-city
  __ngContext__={[Function LRootView]}
  countryService={[Function CountryService]}
>
  <div
    class="container"
  >
    <div
      class="capital-container"
    >
      <div
        class="select-country"
      >
         Select a country 
      </div>
    </div>
  </div>
</app-city>
`;

exports[`ShowCityComponent should display a capital from input country value (Snapshot) 1`] = `
<app-city
  __ngContext__={[Function LRootView]}
  countryService={[Function CountryService]}
>
  <div
    class="container"
  >
    <div
      class="capital-container"
    >
      <div
        class="capital-selected"
      >
        <div
          class="country-name"
        >
          <span>
             Country 
          </span>
          <span>
            Spain
          </span>
        </div>
        <div
          class="city-name"
        >
          <span>
             Capital 
          </span>
          <span>
             Madrid 
          </span>
        </div>
      </div>
    </div>
  </div>
</app-city>
`;

exports[`ShowCityComponent should display a capital from input country value 1`] = `
<app-city
  __ngContext__={[Function LRootView]}
  countryService={[Function CountryService]}
>
  <div
    class="container"
  >
    <div
      class="capital-container"
    >
      <div
        class="capital-selected"
      >
        <div
          class="country-name"
        >
          <span>
             Country 
          </span>
          <span>
            Spain
          </span>
        </div>
        <div
          class="city-name"
        >
          <span>
             Capital 
          </span>
          <span>
             Madrid 
          </span>
        </div>
      </div>
    </div>
  </div>
</app-city>
`;
